src/TimerContext.tsx

```diff
@@ -0,0 +1,211 @@
+    restTime: number;
+    isWorking: boolean;
+    duration: number; // Current interval time remaining
+}
+
+// Union type of all timer types
+export type Timer = StopwatchTimer | CountdownTimer | XYTimer | TabataTimer;
+
+export type TimerContextType = {
+    timers: Timer[];
+    currentTimerIndex: number | null;
+    toggleStartPause: () => void;
+    fastForward: () => void;
+    addTimer: (timer: Timer) => void;
+    removeTimer: (id: string) => void;
+    resetTimers: () => void;
+    getTotalTime: () => number;
+};
+
+// Export the context itself so components can use it directly if needed
+export const TimerContext = createContext<TimerContextType | undefined>(undefined);
+
+export const TimerProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
+    const [timers, setTimers] = useState<Timer[]>([]);
+    const [currentTimerIndex, setCurrentTimerIndex] = useState<number | null>(null);
+
+    useEffect(() => {
+        if (currentTimerIndex === null) return;
+
+        const currentTimer = timers[currentTimerIndex];
+        if (currentTimer?.status !== 'running') return;
+
+        const intervalId = setInterval(() => {
+            setTimers(prevTimers => {
+                const updatedTimers = [...prevTimers];
+                const timer = updatedTimers[currentTimerIndex];
+
+                if (!timer) return prevTimers;
+
+                switch (timer.type) {
+                    case 'stopwatch':
+                        timer.duration += 10;
+                        break;
+
+                    case 'countdown':
+                        timer.duration = Math.max(0, timer.duration - 10);
+                        if (timer.duration === 0) {
+                            timer.status = 'completed';
+                        }
+                        break;
+
+                    case 'XY':
+                    case 'tabata':
+                        timer.duration = Math.max(0, timer.duration - 10);
+                        if (timer.duration === 0) {
+                            if (timer.isWorking) {
+                                timer.isWorking = false;
+                                timer.duration = timer.restTime;
+                            } else {
+                                timer.currentRound++;
+                                if (timer.currentRound > timer.rounds) {
+                                    timer.status = 'completed';
+                                } else {
+                                    timer.isWorking = true;
+                                    timer.duration = timer.workTime;
+                                }
+                            }
+                        }
+                        break;
+                }
+
+                return updatedTimers;
+            });
+        }, 10);
+
+        return () => clearInterval(intervalId);
+    }, [currentTimerIndex, timers]);
+
+    const addTimer = (timer: Timer) => {
+        setTimers(prev => [...prev, timer]);
```

Comment:
I think here's where you should set the timer ID and not externally. Think of it as a database, where the ID gets assigned on insertion rather than externally by clients

Reply:
Implemented, thank you professor.